WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,89 +35,154 @@ private class WeeklyRoundupDataProvider {
 
         for site in sites {
             guard let authToken = site.account?.authToken else {
                 continue
             }
 
-            let wpApi = WordPressComRestApi.defaultApi(oAuthToken: authToken, userAgent: WPUserAgent.wordPress())
-
-            guard let dotComID = site.dotComID?.intValue else {
-                onError(DataRequestError.dotComSiteWithoutDotComID(site))
-                continue
-            }
-
-            let statsServiceRemote = StatsServiceRemoteV2(wordPressComRestApi: wpApi, siteID: dotComID, siteTimezone: site.timeZone)
+            group.enter()
 
-            statsServiceRemote.getData(for: .week, endingOn: periodEndDate, limit: 1) { (timeStats: StatsSummaryTimeIntervalData?, error) in
+            self.fetchStats(for: site, endingOn: periodEndDate, authToken: authToken, in: context) { [weak self] result in
                 defer {
-                    statsProcessed = statsProcessed + 1
-
-                    if statsProcessed == sites.count {
-                        let bestBlogStats = self.filterBest(5, from: blogStats)
-
-                        completion(.success(bestBlogStats))
+                    group.leave()
+                }
+                guard let self else {
+                    return
+                }
+                switch result {
+                case .failure(let error):
+                    self.onError(error)
+                case.success(let stats):
+                    if let stats {
+                        blogStats[site.objectID] = stats
                     }
                 }
+            }
+        }
+
+        group.notify(queue: .global()) {
+            context.perform {
+                let bestBlogStats = self.filterBest(5, from: blogStats)
+                completion(.success(bestBlogStats))
+            }
+        }
+    }
 
+    /// Fetches weekly stats data for a given site.
+    ///
+    /// This function fetches the stats for a single site and passes the result to a completion handler.
+    /// If it encounters any error during fetching, it calls the completion handler with an appropriate error object.
+    /// The completion handler is executed in the same queue as the provided `NSManagedObjectContext`.
+    ///
+    /// **Note:** The completion handler is executed on the context's queue.
+    private func fetchStats(
+        for site: Blog,
+        endingOn periodEndDate: Date,
+        authToken: String,
+        in context: NSManagedObjectContext,
+        completion: @escaping (Result< StatsSummaryData?, Error>) -> Void
+    ) {
+        guard let dotComID = site.dotComID?.intValue else {
+            completion(.failure(DataRequestError.dotComSiteWithoutDotComID(site)))
+            return
+        }
+        let wpApi = WordPressComRestApi.defaultApi(oAuthToken: authToken, userAgent: WPUserAgent.wordPress())
+        let statsServiceRemote = StatsServiceRemoteV2(wordPressComRestApi: wpApi, siteID: dotComID, siteTimezone: site.timeZone)
```

Comment:
I'm aware that it's a common pattern in the app to create a "Remote" instance and dispose it after use, but IMO it would still be beneficial to reuse it if we can? What do you think if we create one "Remote" instance in the `getTopSiteStats` function and pass it to the `fetchStats` calls, so that we can re-use the `URLSession` instance underneath?

Reply:
> so that we can re-use the URLSession instance underneath?

I'm not sure I understand this part, but here is a commit e59d8a4365ef35fcfe439e4642a824e6ec22092f passing the service instance to `fetchStats` method. Let me know if I understood your suggestion correctly.

Reply:
What I wanted to suggest is using one `StatsServiceRemoteV2` instance to make HTTP API calls, instead of creating many instance in the for loop. `StatsServiceRemoteV2` uses `URLSession` underneath. In theory, reusing one `URLSession` instance to make HTTP request to one endpoint is more performant than using multiple instances.

However, I just noticed that `StatsServiceRemoteV2` requires a site id to initialize. We kind of have to create multiple instances per site due to this constraint, unfortunately. Feel free to ignore my previous comment.